# Exploring Ebay Car Sales Data

In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. 

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

## Clean Column Names

- 从以上结果可以看出，数据集有20列，大部分为string；
- 一些列有缺失值，但是都不超过20%；
- 列名以”驼峰式“表示，而不是我们习惯的”蛇形“

我们首先把列名改为蛇形

In [4]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
dict_col_name = {'yearOfRegistration':'registration_year',
                'monthOfRegistration':'registration_month',
                'notRepairedDamage':'unrepaired_damage',
                'dateCreated':'ad_created',
                'dateCrawled':'date_crawled',
                'offerType':'offer_type',
                'vehicleType':'vehicle_type',
                'fuelType':'fuel_type',
                'nrOfPictures':'nr_of_pictures',
                'postalCode':'postal_code',
                'lastSeen':'last_seen'}
autos.rename(dict_col_name,axis = 1, inplace = True)

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

注： `df.describe()`方法的`include='all'`参数返回所有列，默认`include=None`只返回数值类

从上述结果看：

-  `seller`和`offer_type`列几乎只含有一个值（49999/50000），意味着这两列的信息没有意义

-  `price`和`odometer`应该为数值型数据，却储存为了**text**的格式

In [8]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [9]:
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

- Remove any non-numeric characters

- Convert the column to a numeric dtype

In [10]:
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(float)

autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(float)

- rename the column to `odometer_km`

In [11]:
autos.rename({'odometer':'odometer_km'}, inplace = True, axis = 1)

In [12]:
autos['odometer_km'].unique()

array([150000.,  70000.,  50000.,  80000.,  10000.,  30000., 125000.,
        90000.,  20000.,  60000.,   5000., 100000.,  40000.])

## Exploring the Odometer and Price Columns

In [13]:
autos['odometer_km'].unique().shape

(13,)

In [14]:
autos['price'].unique().shape

(2357,)

In [15]:
autos[['odometer_km', 'price']].describe()

odometer_km         price
count   50000.000000  5.000000e+04
mean   125732.700000  9.840044e+03
std     40042.211706  4.811044e+05
min      5000.000000  0.000000e+00
25%    125000.000000  1.100000e+03
50%    150000.000000  2.950000e+03
75%    150000.000000  7.200000e+03
max    150000.000000  1.000000e+08

-  `series.sort_values()`根据值来排序

-  `series.sort_index()`根据索引（行名）排序

In [16]:
autos['price'].value_counts().sort_values(ascending = False)

0.0         1421
500.0        781
1500.0       734
2500.0       643
1200.0       639
1000.0       639
600.0        531
800.0        498
3500.0       498
2000.0       460
999.0        434
750.0        433
900.0        420
650.0        419
850.0        410
700.0        395
4500.0       394
300.0        384
2200.0       382
950.0        379
1100.0       376
1300.0       371
3000.0       365
550.0        356
1800.0       355
5500.0       340
1250.0       335
350.0        335
1600.0       327
1999.0       322
            ... 
27280.0        1
35990.0        1
35950.0        1
6545.0         1
23300.0        1
17989.0        1
39979.0        1
4598.0         1
2070.0         1
38400.0        1
13049.0        1
39911.0        1
15099.0        1
129000.0       1
11960.0        1
1275.0         1
14.0           1
5120.0         1
19970.0        1
13888.0        1
19100.0        1
9989.0         1
1775.0         1
70000.0        1
58700.0        1
1494.0         1
23350.0        1
79933.0       

In [17]:
autos['price'].value_counts().sort_index(ascending = False)

99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
3890000.0        1
1300000.0        1
1234566.0        1
999999.0         2
999990.0         1
350000.0         1
345000.0         1
299000.0         1
295000.0         1
265000.0         1
259000.0         1
250000.0         1
220000.0         1
198000.0         1
197000.0         1
194000.0         1
190000.0         1
180000.0         1
175000.0         1
169999.0         1
169000.0         1
163991.0         1
163500.0         1
155000.0         1
151990.0         1
              ... 
66.0             1
65.0             5
60.0             9
59.0             1
55.0             2
50.0            49
49.0             4
47.0             1
45.0             4
40.0             6
35.0             1
30.0             7
29.0             1
25.0             5
20.0             4
18.0             1
17.0             3
15.0             2
14.0             1
13.0             2
12.0             3
11.0        

In [18]:
autos['odometer_km'].value_counts().sort_index(ascending = False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: odometer_km, dtype: int64

移除价格小于100$, 大于等于100,0000$的行：

In [19]:
autos = autos[autos['price'].between(100,1000000)]

In [20]:
autos['price'].value_counts().sort_index(ascending = False)

999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
194000.0      1
190000.0      1
180000.0      1
175000.0      1
169999.0      1
169000.0      1
163991.0      1
163500.0      1
155000.0      1
151990.0      1
145000.0      1
139997.0      1
137999.0      1
135000.0      1
130000.0      1
129000.0      1
128000.0      1
120000.0      2
           ... 
199.0        41
198.0         1
195.0         2
193.0         1
190.0        16
188.0         1
185.0         1
180.0        35
179.0         1
175.0        12
173.0         1
170.0         7
160.0         8
156.0         2
150.0       224
149.0         7
145.0         2
140.0         9
139.0         1
135.0         1
130.0        15
129.0         1
125.0         8
122.0         1
120.0        39
117.0         1
115.0         2
111.0         2
110.0         3
100.0       134
Name: price, Length: 231

In [22]:
date_crawled = autos['date_crawled'].str[:10]
ad_created = autos['ad_created'].str[:10]
last_seen = autos['last_seen'].str[:10]

参数`normalize=True`可以返回百分比

In [23]:
date_crawled.value_counts(normalize=True, dropna=False)

2016-04-03    0.038609
2016-03-20    0.037800
2016-03-21    0.037220
2016-03-12    0.036909
2016-03-14    0.036660
2016-04-04    0.036536
2016-03-07    0.036059
2016-04-02    0.035602
2016-03-28    0.034960
2016-03-19    0.034732
2016-03-15    0.034317
2016-03-29    0.034130
2016-03-30    0.033736
2016-04-01    0.033695
2016-03-08    0.033176
2016-03-09    0.033011
2016-03-22    0.032886
2016-03-11    0.032596
2016-03-26    0.032306
2016-03-23    0.032285
2016-03-10    0.032285
2016-03-31    0.031849
2016-03-17    0.031518
2016-03-25    0.031497
2016-03-27    0.031124
2016-03-16    0.029465
2016-03-24    0.029444
2016-03-05    0.025359
2016-03-13    0.015676
2016-03-06    0.014038
2016-04-05    0.013063
2016-03-18    0.012897
2016-04-06    0.003172
2016-04-07    0.001389
Name: date_crawled, dtype: float64

In [24]:
ad_created.value_counts(normalize=True, dropna=False)

2016-04-03    0.038858
2016-03-20    0.037863
2016-03-21    0.037448
2016-04-04    0.036888
2016-03-12    0.036743
2016-04-02    0.035291
2016-03-14    0.035291
2016-03-28    0.035063
2016-03-07    0.034794
2016-03-29    0.034089
2016-03-15    0.034047
2016-04-01    0.033674
2016-03-19    0.033612
2016-03-30    0.033550
2016-03-08    0.033176
2016-03-09    0.033093
2016-03-11    0.032907
2016-03-22    0.032700
2016-03-26    0.032368
2016-03-23    0.032119
2016-03-10    0.031995
2016-03-31    0.031891
2016-03-25    0.031621
2016-03-17    0.031186
2016-03-27    0.031041
2016-03-16    0.029962
2016-03-24    0.029382
2016-03-05    0.022912
2016-03-13    0.017044
2016-03-06    0.015303
                ...   
2016-02-26    0.000041
2016-02-14    0.000041
2016-01-10    0.000041
2016-02-05    0.000041
2016-02-24    0.000041
2016-02-02    0.000041
2016-02-20    0.000041
2016-02-12    0.000041
2016-02-18    0.000041
2015-12-05    0.000021
2016-01-29    0.000021
2016-01-22    0.000021
2016-01-14 

In [25]:
last_seen.value_counts(normalize=True, dropna=False)

2016-04-06    0.221971
2016-04-07    0.132146
2016-04-05    0.125054
2016-03-17    0.028096
2016-04-03    0.025131
2016-04-02    0.024882
2016-03-30    0.024696
2016-04-04    0.024530
2016-03-31    0.023825
2016-03-12    0.023783
2016-04-01    0.022850
2016-03-29    0.022311
2016-03-22    0.021357
2016-03-28    0.020860
2016-03-20    0.020652
2016-03-21    0.020549
2016-03-24    0.019761
2016-03-25    0.019097
2016-03-23    0.018579
2016-03-26    0.016671
2016-03-16    0.016443
2016-03-15    0.015862
2016-03-19    0.015759
2016-03-27    0.015551
2016-03-14    0.012628
2016-03-11    0.012400
2016-03-10    0.010637
2016-03-09    0.009580
2016-03-13    0.008875
2016-03-08    0.007320
2016-03-18    0.007320
2016-03-07    0.005433
2016-03-06    0.004313
2016-03-05    0.001078
Name: last_seen, dtype: float64

## Dealing with Incorrect Registration Year Data

In [26]:
autos['registration_year'].describe()

count    48227.000000
mean      2004.730151
std         87.894768
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

可以看到注册年从1000 - 9999年都有，这是不符合常理的。

从之前的`last_seen`可以看出，所有的观测值最后一次被爬取都是在2016年，所以如果一个ad在2016年以后被注册，是不可能的。

以下我们选择`registration_year`在[1990,2016]之间的数据

In [27]:
autos.loc[autos['registration_year'].between(1990,2017),'registration_year'].describe()

count    46461.000000
mean      2004.005854
std          6.108691
min       1990.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2017.000000
Name: registration_year, dtype: float64

In [28]:
autos = autos[autos['registration_year'].between(1990,2017)]

In [29]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46461 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          46461 non-null object
name                  46461 non-null object
seller                46461 non-null object
offer_type            46461 non-null object
price                 46461 non-null float64
abtest                46461 non-null object
vehicle_type          42621 non-null object
registration_year     46461 non-null int64
gearbox               44468 non-null object
powerPS               46461 non-null int64
model                 44332 non-null object
odometer_km           46461 non-null float64
registration_month    46461 non-null int64
fuel_type             42958 non-null object
brand                 46461 non-null object
unrepaired_damage     38140 non-null object
ad_created            46461 non-null object
nr_of_pictures        46461 non-null int64
postal_code           46461 non-null int64
last_seen             46461 non-null object
dtypes: float6

In [30]:
autos['registration_year'].value_counts(normalize=True)

2000    0.066809
2005    0.062655
1999    0.061966
2004    0.058092
2003    0.057963
2006    0.057425
2001    0.056585
2002    0.053314
1998    0.050365
2007    0.048923
2008    0.047567
2009    0.044790
1997    0.041433
2011    0.034825
2010    0.034158
2017    0.029767
1996    0.029164
2012    0.028153
2016    0.025871
1995    0.025677
2013    0.017240
2014    0.014249
1994    0.013474
1993    0.009040
2015    0.008179
1992    0.007899
1991    0.007275
1990    0.007146
Name: registration_year, dtype: float64

## Exploring Price by Brand

In [31]:
brands = autos['brand'].unique()
brands

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'opel', 'mazda',
       'porsche', 'mini', 'toyota', 'dacia', 'nissan', 'jeep', 'saab',
       'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda', 'subaru',
       'sonstige_autos', 'kia', 'citroen', 'chevrolet', 'hyundai',
       'honda', 'daewoo', 'suzuki', 'land_rover', 'alfa_romeo', 'rover',
       'trabant', 'daihatsu', 'lancia', 'lada'], dtype=object)

In [34]:
avg_price_by_brand = {}
for brand in brands:
    rows = autos[autos['brand']==brand]
    avg_price = rows['price'].mean()
    avg_price_by_brand[brand] = avg_price

In [37]:
avg_price_by_brand

{'alfa_romeo': 3440.1040268456377,
 'audi': 9365.683124539426,
 'bmw': 8359.893150684931,
 'chevrolet': 6193.911392405063,
 'chrysler': 3477.6441717791413,
 'citroen': 3658.6854103343467,
 'dacia': 5920.3828125,
 'daewoo': 1105.6756756756756,
 'daihatsu': 1630.8728813559321,
 'fiat': 2707.7714762301916,
 'ford': 3771.2100400369573,
 'honda': 4105.268292682927,
 'hyundai': 5402.913865546218,
 'jaguar': 11494.426470588236,
 'jeep': 11730.621359223302,
 'kia': 5942.905882352941,
 'lada': 2888.409090909091,
 'lancia': 3181.3529411764707,
 'land_rover': 19065.41052631579,
 'mazda': 4100.328275862069,
 'mercedes_benz': 8514.568993319512,
 'mini': 10616.142857142857,
 'mitsubishi': 3428.2077922077924,
 'nissan': 4705.1406896551725,
 'opel': 2961.440567711413,
 'peugeot': 3085.2634560906517,
 'porsche': 48971.62295081967,
 'renault': 2409.066873339238,
 'rover': 1586.4923076923078,
 'saab': 3245.9324324324325,
 'seat': 4388.873873873874,
 'skoda': 6407.966101694915,
 'smart': 3550.305555555555

选取最常见的20个品牌：

In [67]:
top20_common_brands = autos['brand'].value_counts().index[:20]
#如果不加.index，那么选取的是value，而不是brand名

In [68]:
top20_common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'mini', 'volvo', 'mitsubishi'],
      dtype='object')

In [71]:
avg_price_by_20brand = {}
for brand in top20_common_brands:
    rows = autos[autos['brand']==brand]
    avg_price = rows['price'].mean()
    avg_price_by_20brand[brand] = avg_price

In [72]:
avg_price_by_20brand

{'audi': 9365.683124539426,
 'bmw': 8359.893150684931,
 'citroen': 3658.6854103343467,
 'fiat': 2707.7714762301916,
 'ford': 3771.2100400369573,
 'hyundai': 5402.913865546218,
 'mazda': 4100.328275862069,
 'mercedes_benz': 8514.568993319512,
 'mini': 10616.142857142857,
 'mitsubishi': 3428.2077922077924,
 'nissan': 4705.1406896551725,
 'opel': 2961.440567711413,
 'peugeot': 3085.2634560906517,
 'renault': 2409.066873339238,
 'seat': 4388.873873873874,
 'skoda': 6407.966101694915,
 'smart': 3550.3055555555557,
 'toyota': 5151.639398998331,
 'volkswagen': 5379.343702502786,
 'volvo': 4843.0}

## Storing Aggregate Data in a DataFrame

`pd.Series()`构造函数，输入字典，将字典转换为series格式,key变为index，value变为观测值

In [74]:
ap20b = pd.Series(avg_price_by_20brand)
print(ap20b)

audi              9365.683125
bmw               8359.893151
citroen           3658.685410
fiat              2707.771476
ford              3771.210040
hyundai           5402.913866
mazda             4100.328276
mercedes_benz     8514.568993
mini             10616.142857
mitsubishi        3428.207792
nissan            4705.140690
opel              2961.440568
peugeot           3085.263456
renault           2409.066873
seat              4388.873874
skoda             6407.966102
smart             3550.305556
toyota            5151.639399
volkswagen        5379.343703
volvo             4843.000000
dtype: float64


`pd.DataFrame()`构造函数，输入字典、数组、series，返回dataframe

In [84]:
df = pd.DataFrame(ap20b, columns=['mean_price'])
df

mean_price
audi            9365.683125
bmw             8359.893151
citroen         3658.685410
fiat            2707.771476
ford            3771.210040
hyundai         5402.913866
mazda           4100.328276
mercedes_benz   8514.568993
mini           10616.142857
mitsubishi      3428.207792
nissan          4705.140690
opel            2961.440568
peugeot         3085.263456
renault         2409.066873
seat            4388.873874
skoda           6407.966102
smart           3550.305556
toyota          5151.639399
volkswagen      5379.343703
volvo           4843.000000

In [81]:
avg_mileage_by_20brand = {}
for brand in top20_common_brands:
    df = autos[autos['brand']==brand]
    avg_mileage = df['odometer_km'].mean()
    avg_mileage_by_20brand[brand] = avg_mileage
avg_mileage_by_20brand

{'audi': 129349.05428641611,
 'bmw': 132883.56164383562,
 'citroen': 120562.31003039514,
 'fiat': 118356.96413678065,
 'ford': 125275.6390514321,
 'hyundai': 107058.82352941176,
 'mazda': 124841.37931034483,
 'mercedes_benz': 131234.73853950703,
 'mini': 89261.50121065375,
 'mitsubishi': 127207.79220779221,
 'nissan': 118496.55172413793,
 'opel': 130012.81293120442,
 'peugeot': 127330.0283286119,
 'renault': 128523.02922940656,
 'seat': 122015.76576576577,
 'skoda': 111355.93220338984,
 'smart': 100562.86549707603,
 'toyota': 116227.0450751252,
 'volkswagen': 129409.2613233357,
 'volvo': 139890.7766990291}

In [82]:
am20b = pd.Series(avg_mileage_by_20brand)
am20b

audi             129349.054286
bmw              132883.561644
citroen          120562.310030
fiat             118356.964137
ford             125275.639051
hyundai          107058.823529
mazda            124841.379310
mercedes_benz    131234.738540
mini              89261.501211
mitsubishi       127207.792208
nissan           118496.551724
opel             130012.812931
peugeot          127330.028329
renault          128523.029229
seat             122015.765766
skoda            111355.932203
smart            100562.865497
toyota           116227.045075
volkswagen       129409.261323
volvo            139890.776699
dtype: float64

In [85]:
df['mean_mileage'] = am20b
df

mean_price   mean_mileage
audi            9365.683125  129349.054286
bmw             8359.893151  132883.561644
citroen         3658.685410  120562.310030
fiat            2707.771476  118356.964137
ford            3771.210040  125275.639051
hyundai         5402.913866  107058.823529
mazda           4100.328276  124841.379310
mercedes_benz   8514.568993  131234.738540
mini           10616.142857   89261.501211
mitsubishi      3428.207792  127207.792208
nissan          4705.140690  118496.551724
opel            2961.440568  130012.812931
peugeot         3085.263456  127330.028329
renault         2409.066873  128523.029229
seat            4388.873874  122015.765766
skoda           6407.966102  111355.932203
smart           3550.305556  100562.865497
toyota          5151.639399  116227.045075
volkswagen      5379.343703  129409.261323
volvo           4843.000000  139890.776699